### DISCUSSION TITLE
Parallel jobs running concurrently

### DISCUSSION THREAD

**Discussion by CedricHadjian** [OptimalBits/bull#2163] [General]
I came across this package today in the morning and I loved its features. I'm working on a software that should run async tasks in parallel using proxies, i.e each proxy would be a job and these proxies would be working on bots and making separate async calls. What do you recommend I look into? We are also planning on using taskforce.js, what are the main differences between taskforce.js and bull-board?

**Comment by manast** [comment]
Maybe you can get some inspiration and examples on real use cases here: https://blog.taskforce.sh/

**Reply by CedricHadjian** [reply]
Thank you for sharing this, I indeed did benefit from this, but what if it's an array that I need to make multiple processes on? Should I be looping and adding them to queues one by one?

**Reply by CedricHadjian** [reply]
Should I also be using cluster like the example provided [here](https://github.com/OptimalBits/bull/issues/74#issuecomment-69572618) to create a queue process for each proxy?

**Reply by manast** [reply]
Not sure what you mean with proxy, can you provide more context?

**Reply by CedricHadjian** [reply]
Okay so basically, we will be using multiple proxies to make async calls for bots, which are taken from a pool of bots, we want these proxies to all run in parallel, these proxies have a field called "used" that gets added +1 each time an async call gives us success until it reaches 720, the proxies and the bots along with its data like "used" are stored in the database previously.

An example actions would be, which all these proxies would be running in parallel:

So let's say we got 2 proxies that should be running that have 300, and 600 used fields making a separate async request on a pool of 1000 bots

- Proxy of ID 1 that has a used 300:
- - On the first bot, it receives true from the async call, so it adds 1 to its "used" field and becomes 301
- - On the second bot, it receives true from the async call, so it adds 1 to its "used" field and becomes 302
- - On the third bot, it receives false from the async call, so it doesn't add anything
- - ...
- - Process continues until used reaches 720, and then it(proxy of ID) 1 stops

- Proxy of ID 2 that has a used 600:
- - On the first bot, it receives true from the async call, so it adds 1 to its "used" field and becomes 601
- - On the second bot, it receives false from the async call, so it doesn't add anything
- - On the third bot, it receives true from the async call, so it adds 1 to its "used" field and becomes 602
- - ...
- - Process continues until used reaches 720, and then it(proxy of ID) 2 stops

Somewhere along the way OR at the end of each process, we would be making a call to the database to change the used value of the proxies to 720.

I hope I was clear.

**Reply by CedricHadjian** [reply]
@manast I would really appreciate it if I could get some help with what I should be using to achieve what I want, also, if there's a way to receive premium support, I would also appreciate it. Thanks.

**Reply by CedricHadjian** [reply]
@manast This is a good link that demonstrates my request: https://github.com/OptimalBits/bull/issues/1752, where can I find a good guide for it?

**Reply by manast** [reply]
Ok, so one proxy can only be used to make calls to the "bots" one at a time, not in parallel right? And how many proxies would you have in total? a small static number as 2 or a large dynamic number?

**Reply by manast** [reply]
Anyway, one possible solution:
Just add one job to the queue for every proxy, in the job data store the "used" number and the proxy ID, would be 0 when you add the job initially.
Then in the processor use the proxy ID to perform the call with the proper proxy. If the call succeeds increase the used number and add a new job to the queue with the same proxy ID and the new used number. If the used value gets to 720 you just do not add the job to the queue.
With the above scheme you can have as many proxies as you want and as many processors you want running in parallel, still a given proxy ID will only process one job at a given time.

**Reply by CedricHadjian** [reply]
> Ok, so one proxy can only be used to make calls to the "bots" one at a time, not in parallel right?

Right

> And how many proxies would you have in total? a small static number as 2 or a large dynamic number?

A large dynamic number
